src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -213,13 +213,36 @@ private void DisposeCore()
                 _disposed = true;
                 _keepAliveTimer?.Dispose();
                 _stream.Dispose();
-                _inflater?.Dispose();
-                _deflater?.Dispose();
 
                 if (_state < WebSocketState.Aborted)
                 {
                     _state = WebSocketState.Closed;
                 }
+
+                DisposeSafe(_inflater, _receiveMutex);
+                DisposeSafe(_deflater, _sendMutex);
+            }
+        }
+
+        private static void DisposeSafe(IDisposable? resource, AsyncMutex mutex)
+        {
+            if (resource is not null)
+            {
+                Task lockTask = mutex.EnterAsync(CancellationToken.None);
+
+                if (lockTask.IsCompleted)
+                {
+                    resource.Dispose();
```

Comment:
Should we double-check for null here and below?
```suggestion
                    resource?.Dispose();
```

Reply:
It is not needed, because the resource can never become null. We only need to check once at the beginning of the method.

Reply:
Ah, I see, that's what you removed in the first place, right? once assigned, it will now stay assigned